### PR TITLE
Package request: rethinkdb

### PR DESCRIPTION
**Pull request by ifurther**
#3503

it show `* Error: You seem to be using a version of Clang with a bug that may affect Rethink DB. Please upgrade Clang or specify a different compiler.`

### CODE REVIEW

**Comment by buttaface** [comment]
Does this database build with clang on any platform, say linux? The NDK clang is failing a 4 year-old check that they added, rethinkdb/rethinkdb@3c99aac, so I think any clang would probably trigger it.

**Comment by ifurther** [comment]
I don't know, but I just disable it. And it should have so many depend package.

File: packages/rethinkdb/build.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+TERMUX_PKG_HOMEPAGE=https://www.rethinkdb.com/
+TERMUX_PKG_DESCRIPTION="The open-source database for the realtime web. "
+TERMUX_PKG_LICENSE="Apache2.0"
```
Comment (xeffyr) [inline packages/rethinkdb/build.sh:3]: Wrong license name, must be `Apache-2.0`. Arbitrary names are not allowed neither by lint script nor by our hoster (Bintray).

File: packages/rethinkdb/build.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+TERMUX_PKG_SRCURL=https://github.com/rethinkdb/rethinkdb/archive/$TERMUX_PKG_VERSION.tar.gz
+TERMUX_PKG_SHA256=e3749b644af0b230685a32ef31317a9dfebfee5582ded47c444c1c0eaa83712e
+TERMUX_PKG_DEPENDS="ncurses, boost, curl, openssl, libcrypt, python, jemalloc, protobuf,libprotobuf-c, wget, nodejs"
+#TERMUX_PKG_EXTRA_CONFIGURE_ARGS="--allow-fecth CXX=aarch64-linux-android-clang"
+TERMUX_PKG_BUILD_IN_SRC=true
+
+termux_step_configure() {
+        termux_setup_protobuf
+	cd $TERMUX_PKG_BUILDDIR
+	$TERMUX_PKG_SRCDIR/configure \
```
Comment (xeffyr) [inline packages/rethinkdb/build.sh:15]: Any reason to override defaults defined by https://github.com/termux/termux-packages/blob/master/scripts/build/configure/termux_step_configure_autotools.sh ?

Normally, all custom flags should go to `TERMUX_PKG_EXTRA_CONFIGURE_ARGS`.

File: packages/rethinkdb/build.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+                 --allow-fetch  #CXX=aarch64-linux-android-clang \
+                 #PROTOC=$TERMUX_PREFIX/bin/protoc \
+                 --prefix $TERMUX_PREFIX \
+                 --sysconfdir $TERMUX_PREFIX/etc
```
Comment (xeffyr) [inline packages/rethinkdb/build.sh:19]: Here is **syntax error** - add `\` to the end. But as I wrote before, you shouldn't redefine options to `./configure`.

File: packages/rethinkdb/build.sh (hunk 1 above)
Comment (ifurther) [reply packages/rethinkdb/build.sh:3]: OK, I fix it

File: packages/rethinkdb/build.sh (hunk 2 above)
Comment (ifurther) [reply packages/rethinkdb/build.sh:15]: Hi I get `* Error: Unknown option '--disable-dependency-tracking'` for using `TERMUX_PKG_EXTRA_CONFIGURE_ARGS`

File: packages/jemalloc/build.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+TERMUX_PKG_HOMEPAGE=http://jemalloc.net/
+TERMUX_PKG_DESCRIPTION="jemalloc is a general purpose malloc(3) implementation that emphasizes
+fragmentation avoidance and scalable concurrency support."
```
Comment (xeffyr) [inline packages/jemalloc/build.sh:3]: Short one-line description. No hacks for enabling full multiline description.
Ref: https://github.com/termux/termux-packages/wiki/Creating-new-package#table-of-available-package-control-fields

File: packages/jemalloc/build.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+TERMUX_PKG_LICENSE="CUSTOM"
```
Comment (xeffyr) [inline packages/jemalloc/build.sh:4]: `CUSTOM` should be *lowercase*. I won't change build/upload scripts to understand WhAtEVer CAse license names.

`custom` should be used if package uses self-written license. License of jemalloc is `BSD 2-Clause`.

**Comment by xeffyr** [comment]
Don't force-push unless necessary.

File: packages/rethinkdb/build.sh (hunk 2 above)
Comment (xeffyr) [reply packages/rethinkdb/build.sh]: Ok, I have it as well. Will check what can be done.

Comment (xeffyr) [reply packages/rethinkdb/build.sh]: ./configure is not autotools, so should be overridden. Fixed in latest push.

**Comment by xeffyr** [comment]
Fails with
```
src/rdb_protocol/datum.cc:15:10: fatal error: 'boost/detail/endian.hpp' file not found
#include <boost/detail/endian.hpp>
```

Maybe worth to try with fetched `boost`.

**Comment by ifurther** [comment]
Just test this comment.

File: packages/rethinkdb/backtrace.patch [hunk 6]
```diff
@@ -0,0 +1,74 @@
+diff --git a/src/backtrace.cc b/src/backtrace.cc
+index 9ac88eeb29..c89b95d59b 100644
+--- a/src/backtrace.cc
++++ b/src/backtrace.cc
+@@ -7,7 +7,12 @@
+ #include <atomic>
+ #else
+ #include <cxxabi.h>
++#ifdef (__ANDROID__)
```
Comment (xeffyr) [inline packages/rethinkdb/backtrace.patch:9]: Syntax error.

File: packages/rethinkdb/backtrace.patch [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ #include <execinfo.h>
++#endif
+ #include <sys/wait.h>
+ #endif
+ 
+@@ -203,6 +208,55 @@ std::string address_to_line_t::address_to_line(const std::string &executable, co
+ }
+ #endif
+ 
++#ifdef (__ANDROID__)
```
Comment (xeffyr) [inline packages/rethinkdb/backtrace.patch:22]: Again syntax error.

File: packages/rethinkdb/backtrace.patch [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
++        const char* symbol = "";
++
++        Dl_info info;
++        if (dladdr(addr, &info) && info.dli_sname) {
++            symbol = info.dli_sname;
++        }
++
++        os << "  #" << std::setw(2) << idx << ": " << addr << "  " << symbol << "\n";
++    }
++}
```
Comment (xeffyr) [inline packages/rethinkdb/backtrace.patch:68]: Here I see only code copypasted from stackoverflow, without proper understanding what it does and even without reviewing function signatures.

Here are functions which should be present:
```
extern int backtrace (void **__array, int __size) __nonnull ((1));
extern char **backtrace_symbols (void *const *__array, int __size)
extern void backtrace_symbols_fd (void *const *__array, int __size, int __fd)
```

Comment (ifurther) [reply packages/rethinkdb/backtrace.patch:68]: Sorry I doesn't know it. So I should change it?

Comment (xeffyr) [reply packages/rethinkdb/backtrace.patch:68]: > Sorry I doesn't know it.

Then I submit patch to completely turn off backtracing.

Comment (ifurther) [reply packages/rethinkdb/backtrace.patch]: Thank you

**Comment by xeffyr** [comment]
Requires V8 dependency. Unfortunately I can't get it built:
```
  g++ -pthread -m64 -m64  -o /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/mksnapshot -Wl,--start-group /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/mksnapshot/src/mksnapshot.o /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/tools/gyp/libv8_base.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/tools/gyp/libv8_nosnapshot.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/tools/gyp/libv8_libplatform.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/third_party/icu/libicui18n.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/third_party/icu/libicuuc.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/tools/gyp/libv8_libbase.a /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.host/third_party/icu/libicudata.a -Wl,--end-group -lrt
  LD_LIBRARY_PATH=/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/lib.host:/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../tools/gyp; mkdir -p /home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.target/v8_snapshot/geni; "/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/mksnapshot" --log-snapshot-positions --logfile "/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.target/v8_snapshot/geni/snapshot.log" --random-seed 314159265 "/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/out/arm64.release/obj.target/v8_snapshot/geni/snapshot.cc"
Segmentation fault (core dumped)
```

**Comment by buttaface** [comment]
V8 is already built by [the nodejs package, though disabled for x86_64](https://github.com/termux/termux-packages/blob/master/packages/nodejs/build.sh#L20), should be possible for @ifurther to split it off.

**Comment by ifurther** [comment]
buttaface  But the V8 maybe is only for this package

**Comment by xeffyr** [comment]
V8 will require fix for python interpreter. Fetched version still uses outdated python v2 while our docker image sets /usr/bin/python to be v3 which causes syntax errors in gyp.

**Comment by ifurther** [comment]
The rethinkdb's client is still using python2.

**Comment by xeffyr** [comment]
Yes, but '#!/usr/bin/env python` shebangs cause syntax errors when V8 is starting to build. Docker is set to have /usr/bin/python as python3, just like Termux.

**Comment by ifurther** [comment]
I think the version of V8 using python2.

**Comment by xeffyr** [comment]
Patch is needed in order to proceed.

Instead of `make -f Makefile.android aarch64.release`, there should be `make -f Makefile.android arm64.release`.

**Comment by xeffyr** [comment]
Here is a python error:
```
make[2]: Entering directory '/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build'
PYTHONPATH="/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/tools/generate_shim_headers:/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/build::/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build/build/gyp/pylib:" \
GYP_GENERATORS=make \
build/gyp/gyp --generator-output="out" build/all.gyp \
              -Ibuild/standalone.gypi --depth=. \
              -Dv8_target_arch=arm64 \
               -Dtarget_arch=arm64 \
               \
              -S.arm64.release -Dwerror= -Darm_version=6 -Darm_fpu=vfpv2 -Dwant_separate_host_toolset=1
  File "build/gyp/gyp_main.py", line 12
    except ImportError, e:
                      ^
SyntaxError: invalid syntax
make[2]: *** [Makefile:419: out/Makefile.arm64.release] Error 1
make[2]: Leaving directory '/home/builder/.termux-build/rethinkdb/src/build/external/v8_3.30.33.16-patched/build'
```

**Comment by xeffyr** [comment]
V8's Android makefile is *unsuitable* for use with our NDK version which no longer provides GCC toolchain.

**Comment by xeffyr** [comment]
Closing as inactive.
*When submitting a pull request, don't think that maintainers will do 90% of work for you.*

**Pull request closed without merging**